API,Core: Support Conditional Commits

fqaiser94 opened the pull request:
# Context

Adds support for committing changes to an iceberg table based on whether or not a condition is true ***at*** commit time. 
Not before the commit. 
Not after the commit.
**_At_** commit time. 

This is useful in scenarios where users need a robust guard against potential concurrent commits. For example, some use cases require maintaining a monotonically increasing watermark in the snapshot properties. Our recently released iceberg-kafka-connect connector does this however it can only do this on a best-effort basis because Iceberg does not offer any API for expressing conditional commits. As a result, there is a risk of duplicate-file-appends there. This PR would enable closing that loophole. 

For more history/context/usecases, please see the discussion in https://github.com/apache/iceberg/issues/6514.

Incidentally, Delta (the competing table format) offers a similar feature albeit through a much more restricted API: https://github.com/delta-io/delta/blob/master/PROTOCOL.md#transaction-identifiers

# API design

We need to introduce a new API to allow users to declare the conditions under which a commit is allowed to proceed or not. There are two main options here: 
1. Add a new `void commitIf(List<Validation> validations)` method to the `PendingUpdate` interface. 
	1. See latest commit for implementation
	2. This offers a fluent-style API
2. Add a new `void validate(List<Validation> validations)` method to the `PendingUpdate` interface. 
	1. See first commit for implementation
	2. I feel this is simpler but the API is not as fluent as option 1 which some reviewers raised concerns about


# Note

- Please don't be put off by the size of the PR; 90% of it is purely tests.

fqaiser94 reviewed (commented):
@rdblue would love to get some feedback from you (or someone from your team) on this PR. 

Only about 200 lines have changed on the `main` side, the rest is all just `test` side changes. 

What I've done so far: 
- Introduced a new `interface ValidatablePendingUpdate`
- Added an `abstract class BaseValidatablePendingUpdate` that implements the `ValidatablePendingUpdate` interface
- Using the above, I've migrated the following interfaces/classes from `PendingUpdate` to `ValidatablePendingUpdate` so far: 
	- `UpdateProperties`
		- `PropertiesUpdate`
	- `ExpireSnapshots`
		- `RemoveSnapshots`
	- `SnapshotUpdate`
		- `SnapshotProducer` 
			- `BaseOverwriteFiles`  
			- `BaseReplacePartitions`  
			- `BaseRewriteFiles`  
			- `BaseRewriteManifests`  
			- `BaseRowDelta`  
			- `CherryPickOperation` 
			- `FastAppend`  
			- `MergeAppend`  
			- `MergingSnapshotProducer`  
			- `StreamingDelete`
- Also, modified the  `BaseTransaction` class to be able handle `ValidatablePendingUpdates`. 
	- Note: no changes were needed to the `Transaction` interface. 

I can migrate the rest of the `PendingUpdate` implementors as well. 
So far, I haven't found any `PendingUpdate` interface where it doesn't make sense to migrate it to the new `ValidatablePendingInterface`. 
I would love some feedback in the meantime on the current approach.

rdblue commented on api/src/main/java/org/apache/iceberg/BaseValidatablePendingUpdate.java:
```diff
@@ -0,0 +1,58 @@
+        });
+  }
+
+  protected abstract ThisT returnThis();
```
We name this method `self()` in other places.

fqaiser94 replied:
I noticed this. Unfortunately, I can't call it the same thing because it causes conflicts when both methods are in scope. 
Best I can do is call it `returnSelf()`. 
Let me know if you have a better idea, my java skills aren't the greatest :)

rdblue commented on api/src/main/java/org/apache/iceberg/BaseValidatablePendingUpdate.java:
```diff
@@ -0,0 +1,58 @@
+  private final List<Validation> validations = Lists.newArrayList();
+
+  @Override
+  public ThisT addValidation(Predicate<Table> predicate, ValidationException exception) {
```
It seems odd to pass an exception instance. Can we pass a format string and values instead? Basically, it would look like `Precondition.checkArgument` but with a predicate as the first arg rather than a boolean.

fqaiser94 replied:
Ha, I struggled with the `@FormatMethod` annotation when I tried to do this originally. Fixed now.

rdblue commented on api/src/main/java/org/apache/iceberg/BaseValidatablePendingUpdate.java:
```diff
@@ -0,0 +1,58 @@
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+
+abstract class BaseValidatablePendingUpdate<ThisT, T>
+    implements ValidatablePendingUpdate<ThisT, T> {
```
Are there pending updates for which we don't want to be able to validate the table state?

fqaiser94 replied:
When I initially started working on this PR, I wasn't sure if I would be able to implement those methods for every `PendingUpdate`. Now that I understand most of the codebase, I don't see this as an issue anymore. As a result, I've moved the `addValidation` and `validateCurrent` methods inside the `PendingUpdate` interface and implemented them for every `PendingUpdate` as per your suggestion.

rdblue commented on api/src/main/java/org/apache/iceberg/ValidatablePendingUpdate.java:
```diff
@@ -0,0 +1,53 @@
+import java.util.function.Predicate;
+import org.apache.iceberg.exceptions.ValidationException;
+
+public interface ValidatablePendingUpdate<ThisT, T> extends PendingUpdate<T> {
```
I'd probably include these in `PendingUpdate`.

rdblue commented on core/src/main/java/org/apache/iceberg/BaseOverwriteFiles.java:
```diff
@@ -165,4 +165,9 @@ private Expression dataConflictDetectionFilter() {
   }
+
+  @Override
+  protected SnapshotUpdate<OverwriteFiles> returnThis() {
```
This already implement `self()`.

fqaiser94 replied:
This was a little tricky but I have managed to convince Java's type system that `self` and `returnThis` have the same return type, so we no longer need to specify a `returnThis` method for `SnapshotProducer` inheritors and in fact I can rename the `returnThis` method to just `self` without any type conflicts.

fqaiser94 replied:
Changed my mind about this. 
I've changed the`validateCurrentTable` to now return `void` instead of `ThisT`. 
This makes the code changes much simpler as we no longer need this `self` method nor do we need to do any java-type-system-gymnastics.  
The only downside is that method chaining is no longer possible after calling `validateCurrentTable` method which I don't consider to be a big deal.

rdblue replied:
I don't think this is a good idea. This should use the existing `self()` method and use a similar signature as the others configuration methods.

fqaiser94 replied:
Sorry, this is the only comment I'm confused about. 

Are you saying I should change the return type of the the `validateCurrentTable` method from this: 
```
void validateCurrentTable(Predicate<Table> test, String message, Object... args)
```
to this: 
```
ThisT validateCurrentTable(Predicate<Table> test, String message, Object... args)
```
?

I can do that, but this means I have to the change type signature of PendingUpdate from `PendingUpdate<T>` to `PendingUpdate<ThisT, T>`. 
Every interface that extends this interface and every class that implements this interface will have a breaking change. 
Is that acceptable? 

Sample rev-api output:

```
* What went wrong:
Execution failed for task ':iceberg-api:revapi'.
> There were Java public API/ABI breaks reported by revapi:
  
  java.class.noLongerImplementsInterface: Class no longer implements interface 'org.apache.iceberg.PendingUpdate<java.util.Map<java.lang.String, java.lang.String>>'.
  
  old: interface org.apache.iceberg.UpdateProperties
  new: interface org.apache.iceberg.UpdateProperties
  
  SOURCE: BREAKING, BINARY: BREAKING
  
  From old archive: iceberg-api-1.1.0.jar
  From new archive: iceberg-api-1.2.0-SNAPSHOT.jar
  
  If this is an acceptable break that will not harm your users, you can ignore it in future runs like so for:
  
    * Just this break:
        ./gradlew :iceberg-api:revapiAcceptBreak --justification "{why this break is ok}" \
          --code "java.class.noLongerImplementsInterface" \
          --old "interface org.apache.iceberg.UpdateProperties" \
          --new "interface org.apache.iceberg.UpdateProperties"
    * All breaks in this project:
        ./gradlew :iceberg-api:revapiAcceptAllBreaks --justification "{why this break is ok}"
    * All breaks in all projects:
        ./gradlew revapiAcceptAllBreaks --justification "{why this break is ok}"
  ----------------------------------------------------------------------------------------------------
  java.class.nowImplementsInterface: Class now implements interface 'org.apache.iceberg.PendingUpdate<org.apache.iceberg.UpdateProperties, java.util.Map<java.lang.String, java.lang.String>>'.
  
  old: interface org.apache.iceberg.UpdateProperties
  new: interface org.apache.iceberg.UpdateProperties
  
  SOURCE: NON_BREAKING, BINARY: NON_BREAKING
  
  From old archive: iceberg-api-1.1.0.jar
  From new archive: iceberg-api-1.2.0-SNAPSHOT.jar
  
  If this is an acceptable break that will not harm your users, you can ignore it in future runs like so for:
  
    * Just this break:
        ./gradlew :iceberg-api:revapiAcceptBreak --justification "{why this break is ok}" \
          --code "java.class.nowImplementsInterface" \
          --old "interface org.apache.iceberg.UpdateProperties" \
          --new "interface org.apache.iceberg.UpdateProperties"
    * All breaks in this project:
        ./gradlew :iceberg-api:revapiAcceptAllBreaks --justification "{why this break is ok}"
    * All breaks in all projects:
        ./gradlew revapiAcceptAllBreaks --justification "{why this break is ok}"
  ----------------------------------------------------------------------------------------------------
  java.class.superTypeTypeParametersChanged: Super type's type parameters changed from 'org.apache.iceberg.PendingUpdate<java.util.Map<java.lang.String, java.lang.String>>' to 'org.apache.iceberg.PendingUpdate<org.apache.iceberg.UpdateProperties, java.util.Map<java.lang.String, java.lang.String>>'.
  
  old: interface org.apache.iceberg.UpdateProperties
  new: interface org.apache.iceberg.UpdateProperties
  
  SOURCE: POTENTIALLY_BREAKING, BINARY: POTENTIALLY_BREAKING
  
  From old archive: iceberg-api-1.1.0.jar
  From new archive: iceberg-api-1.2.0-SNAPSHOT.jar
  
  If this is an acceptable break that will not harm your users, you can ignore it in future runs like so for:
  
    * Just this break:
        ./gradlew :iceberg-api:revapiAcceptBreak --justification "{why this break is ok}" \
          --code "java.class.superTypeTypeParametersChanged" \
          --old "interface org.apache.iceberg.UpdateProperties" \
          --new "interface org.apache.iceberg.UpdateProperties"
    * All breaks in this project:
        ./gradlew :iceberg-api:revapiAcceptAllBreaks --justification "{why this break is ok}"
    * All breaks in all projects:
        ./gradlew revapiAcceptAllBreaks --justification "{why this break is ok}"
  ----------------------------------------------------------------------------------------------------
  java.class.noLongerImplementsInterface: Class no longer implements interface 'org.apache.iceberg.PendingUpdate<org.apache.iceberg.Schema>'.
  
  old: interface org.apache.iceberg.UpdateSchema
  new: interface org.apache.iceberg.UpdateSchema
  
  SOURCE: BREAKING, BINARY: BREAKING
  
  From old archive: iceberg-api-1.1.0.jar
  From new archive: iceberg-api-1.2.0-SNAPSHOT.jar
  
  If this is an acceptable break that will not harm your users, you can ignore it in future runs like so for:
  
    * Just this break:
        ./gradlew :iceberg-api:revapiAcceptBreak --justification "{why this break is ok}" \
          --code "java.class.noLongerImplementsInterface" \
          --old "interface org.apache.iceberg.UpdateSchema" \
          --new "interface org.apache.iceberg.UpdateSchema"
    * All breaks in this project:
        ./gradlew :iceberg-api:revapiAcceptAllBreaks --justification "{why this break is ok}"
    * All breaks in all projects:
        ./gradlew revapiAcceptAllBreaks --justification "{why this break is ok}"
  ----------------------------------------------------------------------------------------------------
  java.class.nowImplementsInterface: Class now implements interface 'org.apache.iceberg.PendingUpdate<org.apache.iceberg.UpdateSchema, org.apache.iceberg.Schema>'.
  
  old: interface org.apache.iceberg.UpdateSchema
  new: interface org.apache.iceberg.UpdateSchema
  
  SOURCE: NON_BREAKING, BINARY: NON_BREAKING
  
  From old archive: iceberg-api-1.1.0.jar
  From new archive: iceberg-api-1.2.0-SNAPSHOT.jar
  
  If this is an acceptable break that will not harm your users, you can ignore it in future runs like so for:
  
    * Just this break:
        ./gradlew :iceberg-api:revapiAcceptBreak --justification "{why this break is ok}" \
          --code "java.class.nowImplementsInterface" \
          --old "interface org.apache.iceberg.UpdateSchema" \
          --new "interface org.apache.iceberg.UpdateSchema"
    * All breaks in this project:
        ./gradlew :iceberg-api:revapiAcceptAllBreaks --justification "{why this break is ok}"
    * All breaks in all projects:
        ./gradlew revapiAcceptAllBreaks --justification "{why this break is ok}"
  ----------------------------------------------------------------------------------------------------
  java.class.superTypeTypeParametersChanged: Super type's type parameters changed from 'org.apache.iceberg.PendingUpdate<org.apache.iceberg.Schema>' to 'org.apache.iceberg.PendingUpdate<org.apache.iceberg.UpdateSchema, org.apache.iceberg.Schema>'.
  
  old: interface org.apache.iceberg.UpdateSchema
  new: interface org.apache.iceberg.UpdateSchema
  
  SOURCE: POTENTIALLY_BREAKING, BINARY: POTENTIALLY_BREAKING
  
  From old archive: iceberg-api-1.1.0.jar
  From new archive: iceberg-api-1.2.0-SNAPSHOT.jar
  
  If this is an acceptable break that will not harm your users, you can ignore it in future runs like so for:
  
    * Just this break:
        ./gradlew :iceberg-api:revapiAcceptBreak --justification "{why this break is ok}" \
          --code "java.class.superTypeTypeParametersChanged" \
          --old "interface org.apache.iceberg.UpdateSchema" \
          --new "interface org.apache.iceberg.UpdateSchema"
    * All breaks in this project:
        ./gradlew :iceberg-api:revapiAcceptAllBreaks --justification "{why this break is ok}"
    * All breaks in all projects:
        ./gradlew revapiAcceptAllBreaks --justification "{why this break is ok}"
  ----------------------------------------------------------------------------------------------------
```

rdblue commented on core/src/main/java/org/apache/iceberg/NoOpOperation.java:
```diff
@@ -0,0 +1,53 @@
+
+  @Override
+  public void commit() {
+    validateCurrent(SerializableTable.copyOf(new BaseTable(ops, tableName)));
```
It doesn't seem worth adding table name to all of the operation constructors just to construct this `BaseTable`. What about just passing null? The table is only used for validation.

Also, I think it would be better to use a `BaseTable` with an underlying `StaticTableOperations`. You'll need to add a constructor that allows you to pass in the `TableMetadata`.

fqaiser94 replied:
> Also, I think it would be better to use a BaseTable with an underlying StaticTableOperations.

Great point, done. 

> It doesn't seem worth adding table name to all of the operation constructors just to construct this BaseTable. What about just passing null? The table is only used for validation.

Technically, it's possible for a user to access the table name inside a validation context e.g. 
```
.addValidation(
  currentTable -> Objects.equals(currentTable.name(), "abc"), 
  "Table name doesn't match our expected value of `abc`")
```
This validation will obviously fail if we pass in `null`. I realize this is an extremely unlikely use-case,  however since it's pretty easy for us to pass in the correct table name, I'd rather do the "right" thing than pass around `null` values. 

Those are just my thoughts. If you feel strongly about it, just let me know and I can make the change 👍

stevenzwu commented:
@fqaiser94 I added a comment to the issue regarding the motivation use cases: https://github.com/apache/iceberg/issues/6514#issuecomment-1444388008.

fqaiser94 commented:
@stevenzwu sorry, I've [responded](https://github.com/apache/iceberg/issues/6514#issuecomment-1451250830) in the issue now, let's continue the conversation there.

fqaiser94 commented:
All comments have been addressed. 
This is now ready for a second round of reviews.

fqaiser94 commented:
Sorry @nastra @jackye1995, I didn't mean to remove you both as reviewers.
For some reason, the UI won't even let me add you folks back as reviewers! 😵‍💫

rdblue commented on .palantir/revapi.yml:
```diff
@@ -261,6 +266,17 @@ acceptedBreaks:
+      new: "method void org.apache.iceberg.SetStatistics::<init>(java.lang.String,\
+        \ org.apache.iceberg.TableOperations)"
+      justification: "SetStatistics should only be constructed by Iceberg code. Hence\
+        \ the change of constructor params shouldn't affect users."
```
There should not be revapi exceptions. If you add to an interface, then add a default implementation.

rdblue commented on api/src/main/java/org/apache/iceberg/PendingUpdate.java:
```diff
@@ -29,6 +31,24 @@
+   *     will be thrown if the test returns false.
+   */
+  @FormatMethod
+  void validateCurrentTable(Predicate<Table> test, String failMessage);
```
This only has the failure message, not args for that failure message.

fqaiser94 replied:
🤦 Fixed.

rdblue commented on build.gradle:
```diff
@@ -698,6 +698,11 @@ project(':iceberg-dell') {
     implementation project(':iceberg-common')
     implementation project(path: ':iceberg-bundled-guava', configuration: 'shadow')
     compileOnly 'com.emc.ecs:object-client-bundle'
+    compileOnly("org.apache.hadoop:hadoop-common") {
+      exclude group: 'commons-beanutils'
+      exclude group: 'org.apache.avro', module: 'avro'
+      exclude group: 'org.slf4j', module: 'slf4j-log4j12'
+    }
```
Is this still needed? I don't know why these changes would use `SerializableTable` now.

fqaiser94 replied:
You're right, this is no longer needed, removed these changes.

rdblue commented on core/src/main/java/org/apache/iceberg/BasePendingUpdate.java:
```diff
@@ -0,0 +1,56 @@
+  }
+
+  protected final void validateCurrentTableMetadata(String tableName, TableMetadata base) {
+    Table currentTable = new BaseTable(new StaticTableOperations(base), tableName);
```
I'd rather not have all the changes to pass the table name in here. That makes this touch a lot of files needlessly.

fqaiser94 replied:
OK, `tableName` is always `null` now.

fqaiser94 commented on core/src/main/java/org/apache/iceberg/BasePendingUpdate.java:
```diff
@@ -0,0 +1,56 @@
+
+  private static class Validation {
+    private final Predicate<Table> predicate;
+    private final ValidationException exception;
```
Fixed. 

The reason I had to do this originally was because the compiler kept throwing the following warning: 
```
/Users/fq/src/iceberg/core/src/main/java/org/apache/iceberg/BasePendingUpdate.java:55: error: [FormatStringAnnotation] Variables used as format strings that are not local variables must be compile time constants.
```
Unfortunately, I haven't been able to find a way to convince the compiler, so as a compromise I've had to to include a `@SuppressWarnings("FormatStringAnnotation")` to get around this issue.

nastra commented on api/src/main/java/org/apache/iceberg/PendingUpdate.java:
```diff
@@ -41,13 +65,15 @@
    * <p>Once the commit is successful, the updated table will be refreshed.
    *
    * @throws ValidationException If the update cannot be applied to the current table metadata.
+   * @throws UnsupportedOperationException If any predicate supplied through {@link
```
where is this verified?

nastra commented on api/src/main/java/org/apache/iceberg/PendingUpdate.java:
```diff
@@ -29,6 +31,28 @@
+   *     thrown if the test returns false.
+   */
+  @FormatMethod
+  default void validateCurrentTable(Predicate<Table> test, String message, Object... args) {
```
having `validateCurrentTable(...)` leads to an awkward/non-fluent API. What about having `commitIf(Validation)` that would add the given validation and call `commit()` afterwards?

fqaiser94 replied:
This is very close to what I originally proposed ;) 
Although I would prefer if the `commitIf` method accepted a list of validations rather than a single validation as that would allows users to specify a unique message per failed predicate (which can be useful in recovering from the specific `ValidationException`). To be concrete, this is what I would reccommend: 
```
public interface PendingUpdate<T> {
	void commitIf(List<Validation> validations)
}

public class Validation {
  private final Predicate<Table> predicate;
  private final String message;
  private final Object[] args;

  @FormatMethod
  Validation(Predicate<Table> predicate, String message, Object... args) {
    this.predicate = predicate;
    this.message = message;
    this.args = args;
  }

  @SuppressWarnings("FormatStringAnnotation")
  void check(Table currentTable) {
    ValidationException.check(predicate.test(currentTable), message, args);
  }
}
```

You can see a working implemention of this idea in one of my latest (reverted) commits here: https://github.com/apache/iceberg/pull/6513/commits/dd94de91beb28bdfa19128df54e92a929850f2f3
I think it's worth reconsidering in light of some of the recent [comments](https://github.com/apache/iceberg/pull/6513#discussion_r1129872213) as this could be a good compromise. 
However @rdblue reccommended the current API design so I'd love to get his thoughts before making this change.

fqaiser94 replied:
I agree the `commitIf` API is less-awkward/more-fluent. I've made the change now. Fixed.

nastra commented on core/src/main/java/org/apache/iceberg/BasePendingUpdate.java:
```diff
@@ -0,0 +1,59 @@
+  }
+
+  @SuppressWarnings("FormatStringAnnotation")
+  protected final void validateCurrentTableMetadata(TableMetadata base) {
```
why not just name this `validate(..)`? The RevAPI failure you were seeing is because your branch is on quite an old Iceberg version. The method it was conflicting with was removed in https://github.com/apache/iceberg/commit/5a9eb3c20a3867d6ca5ba0d4bea87e1760bab84c

fqaiser94 replied:
We still can't do this even after rebasing on latest master: 
```
Execution failed for task ':iceberg-core:revapi'.
> There were Java public API/ABI breaks reported by revapi:
  
  java.method.visibilityReduced: visibility reduced
  
  old: method void org.apache.iceberg.BaseReplacePartitions::validate(org.apache.iceberg.TableMetadata)
  new: method void org.apache.iceberg.BasePendingUpdate<T>::validate(org.apache.iceberg.TableMetadata) @ org.apache.iceberg.BaseReplacePartitions
  
  SOURCE: BREAKING, BINARY: BREAKING
  
  From old archive: iceberg-core-1.1.0.jar
  From new archive: iceberg-core-1.3.0-SNAPSHOT.jar
  
  If this is an acceptable break that will not harm your users, you can ignore it in future runs like so for:
  
    * Just this break:
        ./gradlew :iceberg-core:revapiAcceptBreak --justification "{why this break is ok}" \
          --code "java.method.visibilityReduced" \
          --old "method void org.apache.iceberg.BaseReplacePartitions::validate(org.apache.iceberg.TableMetadata)" \
          --new "method void org.apache.iceberg.BasePendingUpdate<T>::validate(org.apache.iceberg.TableMetadata) @ org.apache.iceberg.BaseReplacePartitions"
    * All breaks in this project:
        ./gradlew :iceberg-core:revapiAcceptAllBreaks --justification "{why this break is ok}"
    * All breaks in all projects:
        ./gradlew revapiAcceptAllBreaks --justification "{why this break is ok}"
  ----------------------------------------------------------------------------------------------------
  java.method.movedToSuperClass: The method used to be declared in class 'org.apache.iceberg.BaseReplacePartitions' but is now declared in superclass 'org.apache.iceberg.BasePendingUpdate<T extends java.lang.Object>'.
  
  old: method void org.apache.iceberg.BaseReplacePartitions::validate(org.apache.iceberg.TableMetadata)
  new: method void org.apache.iceberg.BasePendingUpdate<T>::validate(org.apache.iceberg.TableMetadata) @ org.apache.iceberg.BaseReplacePartitions
  
  SOURCE: EQUIVALENT, BINARY: EQUIVALENT
  
  From old archive: iceberg-core-1.1.0.jar
  From new archive: iceberg-core-1.3.0-SNAPSHOT.jar
  
  If this is an acceptable break that will not harm your users, you can ignore it in future runs like so for:
  
    * Just this break:
        ./gradlew :iceberg-core:revapiAcceptBreak --justification "{why this break is ok}" \
          --code "java.method.movedToSuperClass" \
          --old "method void org.apache.iceberg.BaseReplacePartitions::validate(org.apache.iceberg.TableMetadata)" \
          --new "method void org.apache.iceberg.BasePendingUpdate<T>::validate(org.apache.iceberg.TableMetadata) @ org.apache.iceberg.BaseReplacePartitions"
    * All breaks in this project:
        ./gradlew :iceberg-core:revapiAcceptAllBreaks --justification "{why this break is ok}"
    * All breaks in all projects:
        ./gradlew revapiAcceptAllBreaks --justification "{why this break is ok}"
  ----------------------------------------------------------------------------------------------------
  java.element.noLongerDeprecated: The element is no longer deprecated.
  
  old: method void org.apache.iceberg.BaseReplacePartitions::validate(org.apache.iceberg.TableMetadata)
  new: method void org.apache.iceberg.BasePendingUpdate<T>::validate(org.apache.iceberg.TableMetadata) @ org.apache.iceberg.BaseReplacePartitions
  
  SOURCE: EQUIVALENT, BINARY: EQUIVALENT
  
  From old archive: iceberg-core-1.1.0.jar
  From new archive: iceberg-core-1.3.0-SNAPSHOT.jar
  
  If this is an acceptable break that will not harm your users, you can ignore it in future runs like so for:
  
    * Just this break:
        ./gradlew :iceberg-core:revapiAcceptBreak --justification "{why this break is ok}" \
          --code "java.element.noLongerDeprecated" \
          --old "method void org.apache.iceberg.BaseReplacePartitions::validate(org.apache.iceberg.TableMetadata)" \
          --new "method void org.apache.iceberg.BasePendingUpdate<T>::validate(org.apache.iceberg.TableMetadata) @ org.apache.iceberg.BaseReplacePartitions"
    * All breaks in this project:
        ./gradlew :iceberg-core:revapiAcceptAllBreaks --justification "{why this break is ok}"
    * All breaks in all projects:
        ./gradlew revapiAcceptAllBreaks --justification "{why this break is ok}"
  ----------------------------------------------------------------------------------------------------
```

fqaiser94 replied:
After rebasing on top of latest master (again), this was indeed possible. Fixed.

nastra commented on core/src/test/java/org/apache/iceberg/TestTransaction.java:
```diff
@@ -771,4 +773,88 @@ public void testSimpleTransactionNotDeletingMetadataOnUnknownSate() throws IOExc
+    Assert.assertEquals("Table should be on version 0", 0, (int) version());
+
+    Transaction txn = table.newTransaction();
+    AssertHelpers.assertThrows(
```
usages of this class are deprecated now. Please use `Assertions.assertThatThrownBy(..).isInstanceOf(..).hasMessage(..)`

fqaiser94 replied:
Fixed.

nastra commented on core/src/test/java/org/apache/iceberg/TestCustomValidations.java:
```diff
@@ -0,0 +1,643 @@
+@RunWith(Parameterized.class)
+public class TestCustomValidations extends TableTestBase {
+
+  @Parameterized.Parameters(name = "formatVersion = {0}")
```
do we really need to test with different format versions?

fqaiser94 replied:
I copied this because I saw other test suites doing this. 
I agree, I don't think there is a particular need to test against different format versions for this suite of tests. 
Fixed.

nastra commented on core/src/test/java/org/apache/iceberg/TestCustomValidations.java:
```diff
@@ -0,0 +1,643 @@
+      }
+    }
+
+    testValidationBehaviours(
```
I understand that we don't want to repeat code, but reading through this test makes it quite difficult to reason about what's exactly being tested and expected (because we're just calling a bunch of lambdas).

I'm curious what others think, but I would vote for having clear and readable than completely de-duplicated code.

fqaiser94 replied:
I appreciate the politely-worded comment but I am going to push back a little here. 

I did not write de-duplicated code for the sake of de-duplicated code here. In general, I agree with having clear and readable tests over deduplicated code and in fact, I originally wrote the tests without any "deduplication" but quickly found that maintaining them was a massive burden. I "deduplicated" the code to ease this and didn't personally feel the readability was compromised that badly. 

I'll leave this as unresolved for. now to see if anyone else finds the tests extremely unreadable, in which case I'll make the requested change.

fqaiser94 replied:
@nastra I've gotten rid of all the lambdas and written separate tests for each case now 😄

fqaiser94 commented:
I took a little break from this PR because reviews were moving a little slowly and it seemed like this feature wasn't considered a high priority. I have since had/seen a couple of conversations with people interested in this feature and affirmed it's value so I'm thinking now might be a good time to try reviving this PR. 

I've rebased the changes on top of latest master and addressed all of the existing comments. Please take a look :)

fqaiser94 commented on core/src/main/java/org/apache/iceberg/SnapshotManager.java:
```diff
@@ -181,4 +182,12 @@ public void commit() {
     }
   }
+
+  @Override
+  public void commitIf(List<Validation> validations) {
+    commitIfRefUpdatesExist();
+    // Add a no-op UpdateProperties to add given validations to transaction
+    transaction.updateProperties().commitIf(validations);
+    commit();
+  }
```
`SnapshotManager` is the only `PendingUpdate` implementation where I have to implement the `commitIf` method "by hand" i.e. I can't just extend `BasePendingUpdate` like all the other implementations. This is because of the way `SnapshotManager` is implemented in terms of `Transaction` which means I don't have access to any base `TableMetadata` to `validate` directly. Instead, I add a conditional, no-op `UpdateProperties` to the underlying transaction which then  validates the current table state as part of the `Transaction` commit process.
